internal/sdkprovider/kafkatopicrepository/repository.go

```diff
@@ -0,0 +1,163 @@
+	errNotFound = aiven.Error{Status: http.StatusNotFound, Message: "Topic not found"}
+	// errAlreadyExists mimics Aiven "conflict" error. Never wrap it, so it can be determined by aiven.IsAlreadyExists
+	errAlreadyExists = aiven.Error{Status: http.StatusConflict, Message: "Topic conflict, already exists"}
+)
+
+const (
+	// defaultV2ListBatchSize the max size of batch to call V2List
+	defaultV2ListBatchSize = 100
+
+	// defaultV2ListRetryDelay V2List caches results, so we retry it by this delay
+	defaultV2ListRetryDelay = 5 * time.Second
+
+	// defaultWorkerCallInterval how often worker should run
+	defaultWorkerCallInterval = time.Second
+	defaultSeenTopicsSize     = 1000
+	defaultSeenServicesSize   = 10
+)
+
+// New returns process singleton Repository
+func New(client topicsClient) Repository {
+	initOnce.Do(func() {
+		singleRep = newRepository(client)
+		go singleRep.worker()
+	})
+	return singleRep
+}
+
+// Repository CRUD interface for topics
+type Repository interface {
+	Create(ctx context.Context, project, service string, req aiven.CreateKafkaTopicRequest) error
+	Read(ctx context.Context, project, service, topic string) (*aiven.KafkaTopic, error)
+	Update(ctx context.Context, project, service, topic string, req aiven.UpdateKafkaTopicRequest) error
+	Delete(ctx context.Context, project, service, topic string) error
+}
+
+// topicsClient interface for unit tests
+type topicsClient interface {
+	List(ctx context.Context, project, service string) ([]*aiven.KafkaListTopic, error)
+	V2List(ctx context.Context, project, service string, topicNames []string) ([]*aiven.KafkaTopic, error)
+	Create(ctx context.Context, project, service string, req aiven.CreateKafkaTopicRequest) error
+	Update(ctx context.Context, project, service, topic string, req aiven.UpdateKafkaTopicRequest) error
+	Delete(ctx context.Context, project, service, topic string) error
+}
+
+func newRepository(client topicsClient) *repository {
+	r := &repository{
+		client:             client,
+		seenTopics:         make(map[string]bool, defaultSeenTopicsSize),
+		seenServices:       make(map[string]bool, defaultSeenServicesSize),
+		v2ListBatchSize:    defaultV2ListBatchSize,
+		v2ListRetryDelay:   defaultV2ListRetryDelay,
+		workerCallInterval: defaultWorkerCallInterval,
+	}
+	return r
+}
+
+// repository implements Repository
+// Handling thousands of topics might be challenging for the API
+// This repository uses retries, rate-limiting, queueing, caching to provide with best speed/durability ratio
+// Must be used as a singleton. See singleRep.
+type repository struct {
+	sync.Mutex
+	client             topicsClient
+	queue              []*request
+	v2ListBatchSize    int
+	v2ListRetryDelay   time.Duration
+	workerCallInterval time.Duration
+
+	// seenTopics stores topic names from v1List and Create()
+	// because v1List might not return fresh topics
+	seenTopics map[string]bool
```

Comment:
`seenTopics` and `seenServices` are critical components of the `repository` type; please consider adding more detailed comments describing their roles